x-pack/osquerybeat/ext/osquery-extension/pkg/amcache/registry/registry.go

```diff
@@ -0,0 +1,249 @@
+	// read the recovered hive contents into memory
+	recoveredHiveContents, err := io.ReadAll(recoveredHive)
+	if err != nil {
+		log.Errorf("failed to read recovered hive: %s", err.Error())
+		return nil, err
+	}
+
+	// return the recovered hive as a registry
+	return regparser.NewRegistry(bytes.NewReader(recoveredHiveContents))
+}
+
+// findTransactionLogs finds all transaction logs for the given hive file.
+// Registry transaction logs are files that record changes to the Windows
+// Registry to prevent corruption and to allow for recovery
+func findTransactionLogs(filePath string, log *logger.Logger) []string {
+	log = ensureLogger(log)
+	logFiles := make([]string, 0)
+
+	// Get the directory and base name of the file
+	dir := filepath.Dir(filePath)
+	baseName := filepath.Base(filePath)
+
+	// Read the directory
+	entries, err := os.ReadDir(dir)
+	if err != nil {
+		log.Errorf("failed to read directory %s: %v", dir, err)
+		return logFiles
+	}
+
+	// Look for files that start with the base name and end with .LOG*
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
+
+		fileName := entry.Name()
+		pattern := fmt.Sprintf("%s.LOG*", baseName)
+		if match, err := filepath.Match(pattern, fileName); err == nil && match {
+			logFiles = append(logFiles, filepath.Join(dir, fileName))
+		}
+	}
+
+	return logFiles
+}
+
+// LoadRegistry loads the registry from the given file path.
+// If transaction logs are found, it will attempt to recover the registry.
+// If no transaction logs are found, it will load the existing registry.
+// The function returns a registry object and an error.
+// The registry object is the recovered registry if recovery was successful,
+// otherwise it is the existing registry.
+func LoadRegistry(filePath string, log *logger.Logger) (registry *regparser.Registry, recovered bool, err error) {
+	log = ensureLogger(log)
+
+	// ensure a path was provided
+	if filePath == "" {
+		log.Errorf("hive file path is empty")
+		return nil, false, fmt.Errorf("hive file path is empty")
+	}
+
+	transactionLogs := findTransactionLogs(filePath, log)
+	if len(transactionLogs) == 0 {
+		log.Infof("no transaction logs found, loading existing registry")
+		registry, err = loadExistingRegistry(filePath, log)
+		if err != nil {
+			log.Errorf("failed to load existing registry: %v", err)
+			return nil, false, err
+		}
+		return registry, false, nil
+	} else {
+		log.Infof("transaction logs found, recovering registry")
+		registry, err := recoverRegistry(filePath, transactionLogs, log)
+		if err != nil {
+			log.Errorf("failed to recover registry: %v", err)
+			// if recovery fails, try to fall back to the existing registry
+			registry, err = loadExistingRegistry(filePath, log)
+			if err != nil {
+				log.Errorf("failed to load existing registry: %v", err)
+				return nil, false, err
+			}
+			return registry, false, nil
+		} else {
+			// if recovery succeeds, return the recovered registry
+			return registry, true, nil
+		}
+	}
```

Comment:
nit: we can remove one level of indentation for readability
```suggestion
	}
    log.Infof("transaction logs found, recovering registry")
    registry, err := recoverRegistry(filePath, transactionLogs, log)
    if err != nil {
	    log.Errorf("failed to recover registry: %v", err)
	    // if recovery fails, try to fall back to the existing registry
	    registry, err = loadExistingRegistry(filePath, log)
	    if err != nil {
		    log.Errorf("failed to load existing registry: %v", err)
		    return nil, false, err
	    }
    }
    // if recovery succeeds, return the recovered registry
    return registry, true, nil
```